design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,190 @@
+- This design does not attempt to implement a 1:1 model of Kubernetes PersistentVolumes and PersistentVolumeClaims.  Some shallow similarities will exist. 
+- This design does not include the Swift interface implementation by Ceph Object.
+- This design does not provide users a means of deleting buckets via the Kubernetes API in order to avoid accidental loss of data.  A Delete operation only deletes the Kubernetes API objects.
+
+## Requirements
+
+1. The Rook-Ceph operator should be the only object store accessor with bucket CREATE privileges.
+1. Users' access key privileges must only have object PUT, GET, and DELETE rights for the bucket they have requested.
+1. When a user creates a CephObjectBucketClaim instance, the operator should create, in the CephObjectBucketClaims' namespace,
+    1. a ConfigMap with the bucket's connection information
+    1. a Secret with the access key with an object PUT/GET/DELETE ACL.
+
+## Users
+
+- Admin: A Kubernetes cluster administrator with RBAC permissions to instantiate and control access to cluster storage. The admin will use existing Rook-Ceph CRDs to create Ceph Object Stores and StorageClasses to make them accessible to users.  
+- User: A Kubernetes cluster user with limited permissions, typically confined to CRUD operations on Kubernetes objects within a single namespace.  The user will request Ceph Object buckets by instantiating CephObjectBucketClaims. 
+
+## Use Cases
+
+### Assumptions
+
+- A running Kubernetes cluster
+- The Rook-Ceph Operator is deployed in namespace `rook-ceph-system`
+- A CephCluster CRD instance created in namespace `rook-ceph-system`
+- A CephObjectStore CRD instance created in namespace `rook-ceph`
+
+**Use Case: Expose an Object Store Endpoint for Bucket Provisioning**
+
+As an admin, I want to expose an existing object store to cluster users so they can begin bucket provisioning via the Kubernetes API._
+
+![admin-actions](./cobc-admin.png)
+
+1. The admin creates a StorageClass with `parameters[serviceName]=OBJECT-STORE-SERVICE` and `parameters[serviceNamespace]=OBJECT-STORE-NAMESPACE`
+
+**Use Case: Provision a Bucket** 
+
+_As a Kubernetes user, I want to leverage the Kubernetes API to create Ceph Object Buckets. I expect to get back the bucket connection information in a Pod-attachable object._
+ 
+![user-actions](./cobc-user.png)
+
+1. The user creates a CephObjectBucketClaim in their namespace.
+1. The operator detects the new ObjectBucketClaim instance.
+    1. The operator derives the object store service via `storageClass.parameters.objectStoreNamespace` + `storageClass.parameters.objectStore`.
+1. If a CephObjectUser for the CephObjectStore does not exist, the operator creates it and waits for the associated Secret.  This will give the operator **bucket** CREATE/GET/DELETE permissions in the object store.  This key is reused later and exists for the lifetime of the object store.
```

Comment:
What if there are > 1 CephOjjectUsers in the same ns as the ObjectStoreService?

Reply:
Should the ceph object store CRD have an "ownerSecret" field added?

Reply:
Hopefully the _wait_ is only this controller's thread.

Reply:
>What if there are > 1 CephObjectUsers in the same ns as the ObjectStoreService? Choose the 1st?
This may get too into implementation but: operator-created COUs should have it annotated in the metadata.  The operator should list a COUs in a namespace and parse their object stores.  If none are found that have been generated by the operator, create one and use that.  If one exists already with the annotation, use it.  This gives us a more consistent runtime environment and a predictable clean up routine.

> Should the ceph object store CRD have an "ownerSecret" field added?
Out of scope here but yes, if it doesn't already

> Hopefully the _wait_ is only this controller's goroutine.
Right, this is a process of the COBC business logic.  Would include a timeout period.